memory_allocator.c

```diff
@@ -0,0 +1,260 @@
+
+    if (current->mb_size >= total_size)
+      return current;
+  }
+  return NULL;
+}
+
+static mem_block_t *try_allocating_in_area(mem_arena_t *ma,
+    size_t requested_size) {
+  mem_block_t *mb = find_entry(&ma->ma_freeblks,
+                               requested_size + sizeof(mem_block_t));
+
+  if (!mb) /* No entry has enough space. */
+    return NULL;
+
+  TAILQ_REMOVE(&ma->ma_freeblks, mb, mb_list);
+  size_t total_size_left = mb->mb_size - requested_size;
+  if (total_size_left > sizeof(mem_block_t)) {
+    mb->mb_size = -requested_size;
+    mem_block_t *new_mb = (mem_block_t *)((char *)mb + requested_size + sizeof(
+                                            mem_block_t));
+    add_free_memory_block(ma, new_mb, total_size_left);
+  }
+  else
+    mb->mb_size = -mb->mb_size;
+
+  return mb;
+}
+
+void *allocate(size_t size, malloc_pool_t *mp, uint16_t flags) {
+  size_t size_aligned;
+  if (size % ALIGNMENT == 0)
+    size_aligned = size;
+  else
+    size_aligned = size + ALIGNMENT - size % ALIGNMENT;
+
+  /* Search for the first area in the list that has enough space. */
+  mem_arena_t *current = NULL;
+  TAILQ_FOREACH(current, &mp->mp_arena, ma_list) {
+    if (current->ma_magic != MAGIC)
+      panic("Memory corruption detected!");
+
+    mem_block_t *mb = try_allocating_in_area(current, size_aligned);
+
+    if (mb) {
+      add_used_memory_block(current, mb);
+
+      if (flags == M_ZERO)
+        memset(mb->mb_data, 0, size);
+
+      return mb->mb_data;
+    }
+  }
+
+  /* Couldn't find any continuous memory with the requested size. */
+  return NULL;
+}
+
+void deallocate(void *addr, malloc_pool_t *mp) {
+  mem_block_t *mb = (mem_block_t *)(((char *)addr) - sizeof(mem_block_t));
+
+  if (mb->mb_magic != MAGIC || mp->mp_magic != MAGIC || mb->mb_size >= 0)
+    panic("Memory corruption detected!");
+
+  mem_arena_t *current = NULL;
+  TAILQ_FOREACH(current, &mp->mp_arena, ma_list) {
+    char *start = ((char *)current) + sizeof(mem_arena_t);
+    if ((char *)addr >= start && (char *)addr < start + current->ma_size) {
+      remove_used_memory_block(current, mb);
+      add_free_memory_block(current, mb, abs(mb->mb_size) + sizeof(mem_block_t));
+    }
+  }
+}
+
+void print_free_blocks(malloc_pool_t *mp) {
+  mem_arena_t *current_arena = NULL;
+  TAILQ_FOREACH(current_arena, &mp->mp_arena, ma_list) {
+    kprintf("Printing free blocks:\n");
+    mem_block_t *current_block = NULL;
+    TAILQ_FOREACH(current_block, &current_arena->ma_freeblks, mb_list) {
+      kprintf("%d\n", current_block->mb_size);
+    }
+
+    kprintf("Printing used blocks:\n");
+    current_block = NULL;
+    TAILQ_FOREACH(current_block, &current_arena->ma_usedblks, mb_list) {
+      kprintf("%d\n", current_block->mb_size);
+    }
+  }
+  kprintf("\n");
+}
+
+void test_memory_allocator() {
+  char text[] = "this is a pool for testing purposes";
+  MALLOC_DEFINE(test_pool, text);
+  TAILQ_INIT(&test_pool->mp_arena);
+
+
+  char array[2100];
```

Comment:
Please allocate memory from buddy systems. Kernel's stack is usually limited in size to one or two pages, thus deep recursion and array allocations are highly disadvised.